### PR TITLE
[WB-46] Implement initial DAO wrapper.

### PR DESCRIPTION
**Pull request by austinkelleher**

### CODE REVIEW

**Comment by austinkelleher** [comment]
Will fix failing tests because of database not existing.

File: dao/Entity.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  prototype: {
+    getId () {
+      return this.getEntityId()
+    },
+
+    setId (id) {
+      this.setEntityId(id)
+    }
+  }
```
Comment (charlieduong94) [inline dao/Entity.js]: Why not just use `id` as the property and doing this proxying? These methods seem pointless. Alternatively, just leave the `entityId` as is and use the built-in `fashion-model` getters.

File: test/integration/queue/dao/createDaoHelper-test.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        user: 'postgres',
+        password: 'postgres',
+        database: 'windbreaker'
+      },
+      debug: true
+    }
+  })
+
+  t.true(daoHelper instanceof FashionKnex)
+})
```
Comment (charlieduong94) [inline test/integration/queue/dao/createDaoHelper-test.js]: This could technically be a unit test.

File: dao/FashionKnex.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        .from(tableName)
+        .where('entityId', id)
+        .limit(1)
+
+      let errors = []
+
+      wrapped = this._modelType.wrap(result[0], errors)
+
+      if (errors.length) {
+        throw new Error(`Error(s) while wrapping model with data "${result[0]}": "${errors}"`)
```
Comment (charlieduong94) [inline dao/FashionKnex.js]: A unit test for this situation would be good.

**Comment by charlieduong94** [comment]
Aside from my comments above and the failing test case, this looks great. Let's try and get this merged in before my PR (#14) since there were some docker changes made.

File: test/integration/queue/dao/FashionKnex-test.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
+require('require-self-ref')
```
Comment (Druotic) [inline test/integration/queue/dao/FashionKnex-test.js]: Why is this test under `test/integration/queue/dao/`? Maybe put it under `test/integration/dao/` instead? (same applies for other integration tests)

File: test/unit/Entity-test.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+require('require-self-ref')
```
Comment (Druotic) [inline test/unit/Entity-test.js]: Thoughts on putting these unit tests under matching directory structures? e.g. `test/unit/dao/Entity-test.js`

File: test/util/TestEntity.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+require('require-self-ref')
```
Comment (Druotic) [inline test/util/TestEntity.js]: Thoughts on putting this file in `test/dao/util/` to keep utils alongside their consuming tests? (assuming we adopt that sort of directory structure)

File: test/integration/queue/dao/FashionKnex-test.js (hunk 4 above)
Comment (austinkelleher) [inline test/integration/queue/dao/FashionKnex-test.js]: This was a mistake actually. Thanks for catching that.

File: test/unit/Entity-test.js (hunk 5 above)
Comment (austinkelleher) [inline test/unit/Entity-test.js]: Good suggestion. Will change.

File: test/util/TestEntity.js (hunk 6 above)
Comment (austinkelleher) [inline test/util/TestEntity.js]: Agreed.

**Comment by austinkelleher** [comment]
After this PR and @charlieduong94 #14 is merged, I will open another PR to cleanup the test folders per @Druotic's suggestion.

File: dao/FashionKnex.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    } = options
+
+    this._logger = conflogger.configure(logger)
+    this._modelType = modelType
+    this._tableName = tableName || _getTableName(modelType)
+    this._knex = knex(knexConfig)
+  }
+
+  async insert (data, returning) {
+    data = _clean(data)
```
Comment (austinkelleher) [inline dao/FashionKnex.js]: This is a bit tricky since fashion-model is fairly different from knex. I think that as long as the the initial table is created with the correct data types and consumers of this API are passing fashion-models instead of raw data, it should be fine because fashion-model should be doing the validation. I do have some ideas for validation, particularly in creating the table using an initial provided model, but I think I should just get to this in a separate PR.

File: test/util/TestEntity.js (hunk 6 above)
Comment (austinkelleher) [inline test/util/TestEntity.js]: Actually I went back on this since it's shared on the integration and unit tests. Perhaps we could move it to `test/util/dao/TestEntity.js`, but I want to do all of the cleanup in another PR. There is some other work I want to do, but this PR is getting large.

File: .travis.yml [hunk 8]
```diff
@@ -6,6 +6,11 @@ node_js:
 services:
   - docker
   - rabbitmq
+  - postgresql
+
+before_script:
```
Comment (austinkelleher) [inline .travis.yml]: @charlieduong94 @Druotic You'll need this to run tests with travis

Comment (charlieduong94) [inline .travis.yml]: My PR moves all tests over to docker containers, so we won't have to worry about this too much.

File: dao/FashionKnex.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        .where('id', id)
+        .limit(1)
+    } catch (err) {
+      this._logger.error(`Error fetching id "${id}" from  table "${tableName}"`, err)
+      throw err
+    }
+
+    let errors = []
+
+    wrapped = this._modelType.wrap(result[0], errors)
```
Comment (Druotic) [inline dao/FashionKnex.js]: Hmm, just noticed this. Is this a common pattern with fashion-model wrapping - does fashion-model push wrap errors into the second arg if it's an array? I'm not seeing it in the docs or the code, but I could be interpreting the code incorrectly. Or - is the intention to extend models in the future to have their own wrap function that does exactly that - push errors into an array? If the latter is the case, I'd argue maybe we should throw an error for consistency rather than push to an array.

File: dao/FashionKnex.js (hunk 7 above)
Comment (charlieduong94) [inline dao/FashionKnex.js]: If we plan on enforcing a schema when the table is created than I'm cool with these changes. It's just that at the moment, it looks like any sort of data can be pushed into the table atm (postgres is schemaless so it would allow it).

**Comment by charlieduong94** [comment]
Feel free to merge whenever.

File: dao/FashionKnex.js (hunk 9 above)
Comment (austinkelleher) [inline dao/FashionKnex.js]: Hey, good question. I'm not a huge fan of the API, but fashion-model does not throw an error if properties have invalid types upon wrapping. If an array is passed as the second argument, fashion-model will push all errors upon wrapping to that array. The idea is that, there may be some instances where you may want to ignore properties that are wrapped with invalid types and continue with the process. Kinda weird, but the code is behaving correctly as I wrote it. It also may not be documented. I would honestly prefer an option to ignore errors instead the errors array. It's something we could consider implementing.

**Comment by charlieduong94** [comment]
[Consider adding caching layer with redis](https://trello.com/c/rHMZrKen/43-consider-adding-caching-layer-with-redis)

**Comment by charlieduong94** [comment]
[Create dao wrapper](https://trello.com/c/9zZE1l2l/46-create-dao-wrapper)
